pytorch_lightning/trainer/trainer.py

```diff
@@ -711,42 +711,90 @@ def run_sanity_check(self, ref_model):
             self.on_sanity_check_end()
             self.running_sanity_check = False
 
-    def test(
+    def validate(
         self,
         model: Optional[LightningModule] = None,
-        test_dataloaders: Optional[Union[DataLoader, List[DataLoader]]] = None,
+        val_dataloaders: Optional[Union[DataLoader, List[DataLoader]]] = None,
         ckpt_path: Optional[str] = 'best',
         verbose: bool = True,
         datamodule: Optional[LightningDataModule] = None,
     ):
         r"""
-
-        Separates from fit to make sure you never run on your test set until you want to.
+        Perform one evaluation epoch over the validation set.
 
         Args:
-            ckpt_path: Either ``best`` or path to the checkpoint you wish to test.
-                If ``None``, use the weights from the last epoch to test. Default to ``best``.
-
+            ckpt_path: Either ``best`` or path to the checkpoint you wish to validate.
+                If ``None``, use the current weights of the model. Default to ``best``.
             datamodule: A instance of :class:`LightningDataModule`.
+            model: The model to evaluate.
+            val_dataloaders: Either a single PyTorch DataLoader or a list of them,
+                specifying validation samples.
+            verbose: If True, prints the validation results.
+
+        Returns:
+            The dictionary with final validation results returned by validation_epoch_end.
+            If validation_epoch_end is not defined, the output is a list of the dictionaries
+            returned by validation_step.
+        """
+        # --------------------
+        # SETUP HOOK
+        # --------------------
+        self.verbose_evaluate = verbose
+
+        self.logger_connector.set_stage("validation")
+
+        # If you supply a datamodule you can't supply val_dataloaders
+        if val_dataloaders and datamodule:
+            raise MisconfigurationException(
+                'You cannot pass val_dataloaders to trainer.validate if you supply a datamodule'
+            )
+
+        # Attach datamodule to get setup/prepare_data added to model before the call to it below
+        self.data_connector.attach_datamodule(model or self.get_model(), datamodule, 'validation')
```

Comment:
Would you mind creating an Enum with 'train', 'test', 'validation' and use it instead.

Reply:
I created one in my refactor PR (https://github.com/PyTorchLightning/pytorch-lightning/pull/4875)

https://github.com/PyTorchLightning/pytorch-lightning/pull/4875/files#diff-8cc348656e5d5c12f8bb6930e41240da015887af602c1fcc4a0190420d4f7588R21-R33.

(placed in pytorch_lightning/trainer/connectors/logger_connector/epoch_result_store.py)

Maybe we could have one placed somewhere more general so it is used everywhere:

Reply:
Isn't `LoggerStages` distinguishing between the internal three stages, `train`, `validation` and `test`? This PR really needs a different enum, which discriminates between the top-level methods used by the users to call into PyTorch Lightning: `fit` (which runs the train and validation stages), `validate` and `test`.

This is mainly because there were differences in how the validation stage needs to behave when inside `fit` or `validate`.

How do you propose to move forward here?